metals/src/main/scala/scala/meta/internal/rename/RenameProvider.scala

```diff
@@ -0,0 +1,342 @@
+  private def fileRenames(
+      isOccurence: (String => Boolean) => Boolean,
+      fileChanges: Set[String],
+      newName: String
+  ): Option[LSPEither[TextDocumentEdit, ResourceOperation]] = {
+    fileChanges
+      .find { file =>
+        isOccurence(str => {
+          (str.desc.isType || str.desc.isTerm) && file.endsWith(
+            str.desc.name.value + ".scala"
+          )
+        })
+      }
+      .map { file =>
+        val newFile =
+          file.replaceAll("/[^/]+\\.scala$", s"/$newName.scala")
+        LSPEither.forRight[TextDocumentEdit, ResourceOperation](
+          new RenameFile(file, newFile)
+        )
+      }
+  }
+
+  private def companionReferences(sym: String): List[ReferencesResult] = {
+    val results = for {
+      companionSymbol <- companion(sym).toIterable
+      loc <- definitionProvider.fromSymbol(companionSymbol).asScala
+    } yield referenceProvider.references(toReferenceParams(loc))
+    results.toList
+  }
+
+  private def companion(sym: String) = {
+    val termOrType = sym.desc match {
+      case Descriptor.Type(name) =>
+        Some(Descriptor.Term(name))
+      case Descriptor.Term(name) =>
+        Some(Descriptor.Type(name))
+      case other =>
+        None
+    }
+
+    termOrType.map(
+      name =>
+        Symbols.Global(
+          sym.owner,
+          name
+        )
+    )
+  }
+
+  private def changeClosedFiles(fileEdits: Map[String, List[TextEdit]]) = {
+    fileEdits.toArray.par.foreach {
+      case (file, changes) =>
+        val path = AbsolutePath(Paths.get(new URI(file)))
+        val text = path.readText
+        val newText = TextEdits.applyEdits(text, changes)
+        path.writeText(newText)
+    }
+  }
+
+  private def implementations(
+      textParams: TextDocumentPositionParams,
+      shouldCheckImplementation: Boolean
+  ) = {
+    if (shouldCheckImplementation) {
+      for (loc <- implementationProvider.implementations(textParams))
+        yield {
+          val locParams = toReferenceParams(loc)
+          referenceProvider.references(locParams)
+        }
+    } else {
+      Nil
+    }
+  }
+
+  private def canRenameSymbol(symbol: String, newName: Option[String]) = {
+    val forbiddenMethods = Set("equals", "hashCode", "unapply", "unary_!", "!")
+    val desc = symbol.desc
+    val name = desc.name.value
+    val isForbidden = forbiddenMethods(name)
+    if (isForbidden) {
+      client.showMessage(forbiddenRename(name, newName))
+    }
+    val colonNotAllowed = name.endsWith(":") && newName.exists(!_.endsWith(":"))
+    if (colonNotAllowed) {
+      client.showMessage(forbiddenColonRename(name, newName))
+    }
+    symbolIsLocal(symbol) && (!desc.isMethod || (!colonNotAllowed && !isForbidden))
+  }
+
+  private def symbolIsLocal(symbol: String): Boolean = {
+
+    def isFromWorkspace =
+      index
+        .definition(MSymbol(symbol))
+        .exists { definition =>
+          workspace.toNIO.getFileSystem == definition.path.toNIO.getFileSystem &&
```

Comment:
Duplicate logic from `isDependencySource` in https://github.com/scalameta/metals/blob/f54ca4951c4f0ba16b87c13d7308f475305ac81f/metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala#L252

Reply:
The second part is negated, so I didn't really see any sensible way to reuse it. I will move ti the extension method and extract the two conditions.